components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.rest.api.user.recovery.v2/src/main/java/org/wso2/carbon/identity/rest/api/user/recovery/v2/impl/core/UsernameRecoveryService.java

```diff
@@ -0,0 +1,239 @@
+import org.wso2.carbon.identity.recovery.IdentityRecoveryException;
+import org.wso2.carbon.identity.recovery.dto.RecoveryChannelInfoDTO;
+import org.wso2.carbon.identity.recovery.dto.RecoveryInformationDTO;
+import org.wso2.carbon.identity.recovery.dto.UsernameRecoverDTO;
+
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.impl.core.utils.RecoveryUtil;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.APICall;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.AccountRecoveryType;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.InitRequest;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryChannel;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryChannelInformation;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryRequest;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.UsernameRecoveryNotifyResponse;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+
+/**
+ * Call internal OSGI services to perform username recovery.
+ */
+public class UsernameRecoveryService {
+
+    private static final Log log = LogFactory.getLog(UsernameRecoveryService.class);
+
+    /**
+     * Initiate userName recovery from POST.
+     *
+     * @param initRequest {@link InitRequest} object which holds the information in the account recovery request
+     * @return Response
+     */
+    public Response initiateUsernameRecovery(InitRequest initRequest) {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        Map<String, String> userClaims = RecoveryUtil.buildUserClaimsMap(initRequest.getClaims());
+        try {
+            // Get username recovery notification information.
+            RecoveryInformationDTO recoveryInformationDTO =
+                    UserAccountRecoveryServiceDataHolder.getUsernameRecoveryManager().initiate(userClaims, tenantDomain,
+                            RecoveryUtil.buildPropertiesMap(initRequest.getProperties()));
+            if (recoveryInformationDTO == null) {
+                // If RecoveryChannelInfoDTO is null throw not found error.
+                if (log.isDebugEnabled()) {
+                    String message = "No recovery information username recovery request";
+                    log.debug(message);
+                }
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+            }
+            return Response.ok().entity(buildUsernameRecoveryInitResponse(recoveryInformationDTO, tenantDomain))
+                    .build();
+        } catch (IdentityRecoveryClientException e) {
+            if (StringUtils.isEmpty(e.getErrorCode())) {
+                throw RecoveryUtil.handleException(e.getErrorCode(), Constants.STATUS_CONFLICT_MESSAGE_DEFAULT,
```

Comment:
why do we decide the error as a CONFLICT if the error code is empty?

Reply:
Changed to `INTERNAL_SERVER_ERROR` with 8a5d4f0274d4bc026253b69400b0714ad6acf7a7